pkg/integration/helpers.go

```diff
@@ -0,0 +1,343 @@
+	}
+}
+
+// NewAdminAPIServer runs an Admin API Server and returns a corresponding client.
+func (s *Suite) NewAdminAPIServer(ctx context.Context, tb testing.TB) *AdminClient {
+	srv := s.newAdminAPIServer(ctx, tb)
+	s.adminSrv = srv
+	return NewAdminClient(srv.Addr(), s.adminKey)
+}
+
+// NewAPIServer runs an API Server and returns a corresponding client.
+func (s *Suite) NewAPIServer(ctx context.Context, tb testing.TB) *APIClient {
+	srv := s.newAPIServer(ctx, tb)
+	s.apiSrv = srv
+	return NewAPIClient(srv.Addr(), s.deviceKey)
+}
+
+func (s *Suite) newAdminAPIServer(ctx context.Context, tb testing.TB) *server.Server {
+	// Create the router
+	adminRouter := mux.NewRouter()
+	// Install common security headers
+	adminRouter.Use(middleware.SecureHeaders(ctx, s.cfg.AdminAPISrvConfig.DevMode, "json"))
+
+	// Enable debug headers
+	processDebug := middleware.ProcessDebug(ctx)
+	adminRouter.Use(processDebug)
+
+	// Create the renderer
+	h, err := render.New(ctx, "", s.cfg.APISrvConfig.DevMode)
+	if err != nil {
+		tb.Errorf("failed to create the renderer %v", err)
+	}
+
+	// Setup cacher
+	cacher, err := cache.CacherFor(ctx, &s.cfg.APISrvConfig.Cache, cache.MultiKeyFunc(
+		cache.HMACKeyFunc(sha1.New, s.cfg.APISrvConfig.Cache.HMACKey),
+		cache.PrefixKeyFunc("apiserver:cache:"),
+	))
+	if err != nil {
+		tb.Fatalf("failed to create cacher: %v", err)
+	}
+	tb.Cleanup(func() {
+		if err := cacher.Close(); err != nil {
+			tb.Errorf("got err when cleanup: %v", err)
+		}
+	})
+
+	// Create LimitStore
+	limiterStore, err := ratelimit.RateLimiterFor(ctx, &s.cfg.AdminAPISrvConfig.RateLimit)
+	if err != nil {
+		tb.Errorf("failed to create the limit store %v", err)
+	}
+
+	adminRouter.Handle("/health", controller.HandleHealthz(ctx, &s.cfg.AdminAPISrvConfig.Database, h)).Methods("GET")
+
+	{
+		sub := adminRouter.PathPrefix("/api").Subrouter()
+
+		// Setup API auth
+		requireAPIKey := middleware.RequireAPIKey(ctx, cacher, s.db, h, []database.APIUserType{
+			database.APIUserTypeAdmin,
+		})
+		// Install the APIKey Auth Middleware
+		sub.Use(requireAPIKey)
+
+		issueapiController, err := issueapi.New(ctx, &s.cfg.AdminAPISrvConfig, s.db, limiterStore, h)
+		if err != nil {
+			tb.Errorf("issueapi.New: %w", err)
+		}
+		sub.Handle("/issue", issueapiController.HandleIssue()).Methods("POST")
+
+		codeStatusController := codestatus.NewAPI(ctx, &s.cfg.AdminAPISrvConfig, s.db, h)
+		sub.Handle("/checkcodestatus", codeStatusController.HandleCheckCodeStatus()).Methods("POST")
+		sub.Handle("/expirecode", codeStatusController.HandleExpireAPI()).Methods("POST")
+	}
+
+	srv, err := server.New(s.cfg.AdminAPISrvConfig.Port)
+	if err != nil {
+		tb.Errorf("failed to create server: %w", err)
+	}
+
+	// Stop the server on cleanup
+	stopCtx, stop := context.WithCancel(ctx)
+	tb.Cleanup(stop)
+
+	go func() {
+		if err := srv.ServeHTTPHandler(stopCtx, handlers.CombinedLoggingHandler(os.Stdout, adminRouter)); err != nil {
+			tb.Error(err)
+		}
+	}()
+	return srv
```

Comment:
It doesn't look like this server is ever gracefully stopped or terminated.

Reply:
I think the server is terminated when the provided context is cancelled. The corresponding `stop` function is registered to tb.Cleanup.